File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/degen/DegenWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.degen

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.alert.AlertWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.cultproposals.CultProposalsWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.alert.AlertWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapWireframeContext
import com.sonsofcrypto.web3walletcore.modules.degen.DefaultDegenInteractor
import com.sonsofcrypto.web3walletcore.modules.degen.DefaultDegenPresenter
import com.sonsofcrypto.web3walletcore.modules.degen.DegenWireframe
import com.sonsofcrypto.web3walletcore.modules.degen.DegenWireframeDestination
import com.sonsofcrypto.web3walletcore.services.degen.DegenService

class DefaultDegenWireframe(
    private val parent: WeakRef<Fragment>?,
    private val degenService: DegenService,
    private val networksService: NetworksService,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val cultProposalsWireframeFactory: CultProposalsWireframeFactory,
    private val alertWireframeFactory: AlertWireframeFactory
): DegenWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment)
    }

    override fun navigate(destination: DegenWireframeDestination) {
        when (destination) {
            is DegenWireframeDestination.Swap -> {
                val context = CurrencySwapWireframeContext(
                    Network.ethereum(), Currency.ethereum(), Currency.usdt()
                )
                currencySwapWireframeFactory.make(parent?.get(), context).present()
            }
            is DegenWireframeDestination.Cult -> {
                cultProposalsWireframeFactory.make(parent?.get()).present()
            }
            is DegenWireframeDestination.ComingSoon -> {
                val context = AlertWireframeContext.underConstructionAlert()
                alertWireframeFactory.make(parent?.get(), context).present()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = DegenFragment()
        val interactor = DefaultDegenInteractor(
            degenService,
            networksService,
        )
        val presenter = DefaultDegenPresenter(
            WeakRef(view),
            this,
            interactor
        )
        view.presenter = presenter
        return NavigationFragment(view)
    }
}

File: iosApp/iosApp/Modules/Degen/DegenWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultDegenWireframe {
    private weak var parent: UIViewController?
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let cultProposalsWireframeFactory: CultProposalsWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let degenService: DegenService
    private let networksService: NetworksService

    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        cultProposalsWireframeFactory: CultProposalsWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        degenService: DegenService,
        networksService: NetworksService
    ) {
        
        self.parent = parent
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.cultProposalsWireframeFactory = cultProposalsWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.degenService = degenService
        self.networksService = networksService
    }
}

extension DefaultDegenWireframe {

    func present() {
        let vc = wireUp()
        if let tabVc = parent as? UITabBarController {
            let vcs = (tabVc.viewControllers ?? []) + [vc]
            tabVc.setViewControllers(vcs, animated: false)
        } else {
            parent?.show(vc, sender: self)
        }
    }

    func navigate(to destination: DegenWireframeDestination) {
        if destination is DegenWireframeDestination.Swap {
            guard let network = networksService.network else { return }
            let context = CurrencySwapWireframeContext(
                network: network, currencyFrom: nil, currencyTo: nil
            )
            currencySwapWireframeFactory.make(vc, context: context).present()
        }
        if destination is DegenWireframeDestination.Cult {
            cultProposalsWireframeFactory.make(vc).present()
        }
        if destination is DegenWireframeDestination.ComingSoon {
            alertWireframeFactory.make(vc, context: .underConstructionAlert()).present()
        }
    }
}

private extension DefaultDegenWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultDegenInteractor(
            degenService: degenService,
            networksService: networksService
        )
        let vc: DegenViewController = UIStoryboard(.degen).instantiate()
        let presenter = DefaultDegenPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        nc.tabBarItem = UITabBarItem(
            title: Localized("degen"),
            image: UIImage(named: "tab_icon_degen"),
            tag: 1
        )
        self.vc = nc
        return nc
    }
}
